core/src/main/scala/org/http4s/util/encoding/UriCodingUtils.scala

```diff
@@ -0,0 +1,271 @@
+    // We are decoding, but the encoding procedure is as follows:
+    // replace ' ' with '+'
+    // do not encode * - . 0-9 A-Z _ a-z
+    // do percent-encode everything else
+
+    def decode(s: String) = percentDecode(PctEncoded(s.replace("+", " ")))
+
+    if (encoded.isEmpty)
+      Vector("" -> None)
+    else
+      encoded.split("&", Integer.MAX_VALUE).toVector.map {
+        string =>
+          val (name, value) =
+            string.split("=", 2) match {
+              case Array(name)        => (name, None)
+              case Array(name, value) => (name, Some(value))
+            }
+          (decode(name), value.map(decode))
+      }
+
+  }
+
+  def split[C[_]<:SeqLike[_, C[A]],A](payload: C[A], delimiter: A): Vector[C[A]] = {
+    @tailrec def loop(mark: Int, pos: Int, acc: Vector[C[A]]): Vector[C[A]] = {
+      if (pos < payload.size)
+        if (payload(pos) == delimiter)
+          loop(pos+1, pos+1, acc :+ payload.slice(mark, pos))
+        else
+          loop(mark, pos+1, acc)
+      else acc :+ payload.slice(mark, pos)
+    }
+    loop(0, 0, Vector())
+  }
+
+  def split[@specialized(Byte, Char, Int) A](payload: Array[A], delimiter: A): Vector[Array[A]] = {
+    @tailrec def loop(mark: Int, pos: Int, acc: Vector[Array[A]]): Vector[Array[A]] = {
+      if (pos < payload.length)
+        if (payload(pos) == delimiter)
+          loop(pos+1, pos+1, acc :+ payload.slice(mark, pos))
+        else
+          loop(mark, pos+1, acc)
+      else acc :+ payload.slice(mark, pos)
+    }
+    loop(0, 0, Vector())
+  }
+
+//  // doesn't support "isindex"
+//  // does support charset!
+//  def w3cHtml5FormUrlDecode(payload: Array[Byte], defaultEncoding: Charset = Utf8): Vector[(String, Option[String])] = {
+//    def decode(s: String) = percentDecode(PercentEncodedString(s.replace("+", " ")))
+//
+//    if (payload.isEmpty) Vector()
+//    else {
+//      val kvPairs: Vector[(Array[Byte], Option[Array[Byte]])] =
+//        split(payload, '&'.toByte).map {
+//          kv =>
+//            val delimIndex = kv.indexWhere(_ == '='.toByte)
+//            if (delimIndex < 0) kv -> None
+//            else kv.slice(0, delimIndex) -> Some(kv.slice(delimIndex+1, kv.length))
+//        }
+//      val charset =
+//        kvPairs
+//          .find(kv => kv._1 == "_charset_".getBytes(Utf8))
+//          .flatMap {
+//            kv =>
+//              kv._2
+//                .map(new String(_, Utf8))
+//                .flatMap {
+//                  c => if (Charset.isSupported(c)) Some(Charset.forName(c)) else None
+//                }
+//          }.getOrElse(defaultEncoding)
+//
+//      kvPairs.map {
+//        case (key, value) => decode(new String(key, charset)) -> value.map(new String(_, charset)).map(decode)
+//      }
+//    }
+//  }
+  // commented out because I still don't understand how charset works here, in conjunction with assuming proper input
+  // to PercentEncodedString
+}
```

Comment:
I remember needing some time to wrap my head around this the first time as well. The process basically boils down to:
1. The raw string we get are bytes encoded as ISO88591 with some % encoding. They must be decoded to bytes.
2. There may be a special member which is the charset identifier, if so we use it for the charset, else UTF8.
3. The bytes are then decoded using the charset from step 2.

I think your concern is the first part where we have bytes encoded as chars. The full decode process is [char] -> [byte] -> [char], so the first and last are actually well represented by a `String`.

For this method I'd take in an `encodedChars: String` and omit the charset: it is either UTF-8 or specified by the body. The `percentDecode` function should probably take a charset parameter so it can be used by both form or uri variants. The origional version also had `spaceIsPlus` but it appears you've done away with that by replacing `+` with ` ` (space char) prior to decoding which works but makes an extra `String` in the process.

I'm interested in hearing the motivation for discarding/heavily modifying the code that was in `UrlCodingUtils`: was it incorrect in some way?